packages/@aws-cdk/aws-appsync/lib/private/schema-implementation.ts

```diff
@@ -0,0 +1,322 @@
+
+/**
+ * Utility enum for Schema class
+ */
+enum SchemaMode {
+  FILE = 'FILE',
+  CODE = 'CODE',
+  S3 = 'S3',
+};
+
+/**
+ * The implmentation for the abstract Schema class.
+ *
+ * @internal
+ */
+export class SchemaImpl {
+  /**
+   * The definition for this schema
+   */
+  public readonly definition: string;
+
+  /**
+   * The underlying CFN GraphQL Schema Resource. This
+   * resource is created on `bind` which allows the Schema
+   * to be created outside the CDK (a feature that works well
+   * with the code-first approach).
+   *
+   * @internal
+   */
+  protected _schema!: CfnGraphQLSchema;
+
+  /**
+   * The SchemaMode utilized by the Schema Implementation
+   *
+   * @internal
+   */
+  protected _mode: SchemaMode;
+
+  public constructor() {
+    this.definition = '';
+    this._mode = SchemaMode.FILE;
+  }
+
+  /**
+   * Called when the GraphQL Api is initialized to allow this object to bind
+   * to the stack.
+   */
+  public bind(api: GraphqlApi): CfnGraphQLSchema {
+    if (!this._schema) {
+      this._schema = new CfnGraphQLSchema(api, 'Schema', {
+        apiId: api.apiId,
+        definition: this.definition,
+      });
+    }
+    return this._schema;
+  }
+
+  /**
+   * Escape hatch to add to Schema as desired. Will always result
+   * in a newline.
+   */
+  public addToSchema(_addition: string, _delimiter?: string): void {
+    throw new Error('API cannot append to schema because schema definition mode is not configured as CODE.');
+  }
+
+  /**
+   * Add a query field to the schema's Query. CDK will create an
+   * Object Type called 'Query'. For example,
+   *
+   * type Query {
+   *   fieldName: Field.returnType
+   * }
+   */
+  public addQuery(_fieldName: string, _field: ResolvableField): ObjectType {
+    throw new Error(`Unable to add query. Schema definition mode must be ${SchemaMode.CODE}. Received: ${this._mode}`);
+  }
+
+  /**
+   * Add a mutation field to the schema's Mutation. CDK will create an
+   * Object Type called 'Mutation'. For example,
+   *
+   * type Mutation {
+   *   fieldName: Field.returnType
+   * }
+   */
+  public addMutation(_fieldName: string, _field: ResolvableField): ObjectType {
+    throw new Error(`Unable to add mutation. Schema definition mode must be ${SchemaMode.CODE}. Received: ${this._mode}`);
+  }
+
+  /**
+   * Add a subscription field to the schema's Subscription. CDK will create an
+   * Object Type called 'Subscription'. For example,
+   *
+   * type Subscription {
+   *   fieldName: Field.returnType
+   * }
+   */
+  public addSubscription(_fieldName: string, _field: ResolvableField): ObjectType {
+    throw new Error(`Unable to add subscription. Schema definition mode must be ${SchemaMode.CODE}. Received: ${this._mode}`);
+  }
+
+  /**
+   * Add type to the schema
+   */
+  public addType(_type: IIntermediateType): IIntermediateType {
+    throw new Error('API cannot add type because schema definition mode is not configured as CODE.');
+  }
```

Comment:
I believe you're doing this backwards.

All these methods should be `asbtract` in `Schema`, and then `SchemaImpl` should actually implement them.